### ISSUE TITLE
Question about "D_noise_loss"

### ISSUE THREAD

**Issue by diegovalenzuelaiturra** [hehaodele/ProbGAN#2]
Hi, really nice paper and implementation :)

I was looking at the code and this part was not clear to me. I must say I am not familiar with "Stochastic Gradient Hamiltonian Monte Carlo" so the first thing I wanted to know is if it's related to it. 

https://github.com/hehaodele/ProbGAN/blob/e60ab2b867a349e182a643f9e33d1ff04560d97d/train.py#L164

Also, I would appreciate if you could explain to me the idea behind this formulation.

**Comment by hehaodele** [comment]
Hi Diego,

It is right that  D_noise_loss is from "SGHMC".
SGHMC requires adding Gaussian noise to the neural network weights.

An elegant way to implement this is to create a virtual noise_loss = gaussian * weight. 
Thus by taking the gradient of the "noise_loss", you get the gaussian noise.
Of course, you need to sample gaussian noise every time you compute the noise_loss, which is exactly what D_noise_loss is doing.
